src/webgpu/shader/execution/expression/call/builtin/atomics/atomicCompareExchangeWeak.spec.ts

```diff
@@ -34,6 +177,337 @@ struct __atomic_compare_exchange_result<T> {
+      outputBuffer.unmap();
+    }
+
+    const exchangedBuffer = t.device.createBuffer({
+      size: bufferNumElements * arrayType.BYTES_PER_ELEMENT,
+      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
+      mappedAtCreation: true,
+    });
+    t.trackForCleanup(exchangedBuffer);
+    {
+      const data = new arrayType(exchangedBuffer.getMappedRange());
+      data.forEach((_, i) => (data[i] = defaultValue));
+      exchangedBuffer.unmap();
+    }
+
+    const bindGroup = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: [
+        { binding: 0, resource: { buffer: inputBuffer } },
+        { binding: 1, resource: { buffer: outputBuffer } },
+        { binding: 2, resource: { buffer: exchangedBuffer } },
+      ],
+    });
+
+    // Run the shader.
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatchWorkgroups(t.params.dispatchSize);
+    pass.end();
+    t.queue.submit([encoder.finish()]);
+
+    // Read back buffers
+    const outputBufferResult = (
+      await t.readGPUBufferRangeTyped(outputBuffer, {
+        type: arrayType,
+        typedLength: outputBuffer.size / arrayType.BYTES_PER_ELEMENT,
+      })
+    ).data;
+    const exchangedBufferResult = (
+      await t.readGPUBufferRangeTyped(exchangedBuffer, {
+        type: arrayType,
+        typedLength: exchangedBuffer.size / arrayType.BYTES_PER_ELEMENT,
+      })
+    ).data;
+
+    let nextExpectedValue = 0;
+    for (let w = 0; w < numWrites; ++w) {
+      const offset = w * numInvocations;
+      const exchanged = exchangedBufferResult.subarray(offset, offset + numInvocations);
+      const output = outputBufferResult.subarray(offset, offset + numInvocations);
+
+      const dumpValues = () => {
+        return `\n\tFor write: ${w}
+        \texchanged: ${exchanged}
+        \toutput:    ${output}`;
+      };
+
+      // Only one of the invocations should have succeeded to exchange - or none if a spurious failure occured
+      if (exchanged.every(v => v === 0)) {
+        // Spurious failure, all values in output be the default value
+        if (!output.every(v => v === defaultValue)) {
+          t.fail(
+            `Spurious failure detected, expected only default value of ${defaultValue} in output buffer.${dumpValues()}`
+          );
+          return;
+        }
+      } else {
+        // Only one invocation should have exchanged its value
+        if (exchanged.filter(v => v === 1).length !== 1) {
+          t.fail(`More than one invocation exchanged its value.${dumpValues()}`);
+          return;
+        }
+
+        // Get its index
+        const idx = exchanged.findIndex(v => v === 1);
+        assert(idx !== -1);
+
+        // Output should contain the old value after exchange
+        if (output[idx] !== nextExpectedValue) {
```

Comment:
All invocations should have the same old value, no?

Reply:
It didn't, as this version was only updating the output buffer for the invocation that succeeded at exchanging. But I've since updated the code so that we always write the returned value to output, and test that it contains what we expect (that is: the old value for the invocation that exchanged, else the old or new value for all other invocations).